lib/fss/s3.ex

```diff
@@ -89,23 +114,28 @@ defmodule FSS.S3 do
         config =
           opts
           |> Keyword.fetch!(:config)
-          |> case do
-            nil ->
-              config_from_system_env()
-
-            %Config{} = config ->
-              config
-
-            config when is_list(config) or is_map(config) ->
-              struct!(config_from_system_env(), config)
-
-            other ->
-              raise ArgumentError,
-                    "expect configuration to be a %FSS.S3.Config{} struct, a keyword list or a map. Instead got #{inspect(other)}"
-          end
+          |> normalize_config!()
           |> validate_config!()
+          |> then(fn %Config{} = config ->
+            config = %Config{config | bucket: bucket}
+
+            if is_nil(config.endpoint) do
+              # We only use the path-style if the bucket name contain dots.
+              # The standard way is to use the virtual-host style.
+              endpoint =
+                if String.contains?(bucket, ".") do
+                  "https://s3." <> config.region <> ".amazonaws.com/" <> bucket
+                else
+                  "https://" <> bucket <> ".s3." <> config.region <> ".amazonaws.com"
+                end
+
+              %Config{config | endpoint: endpoint}
+            else
+              config
+            end
```

Comment:
Given my comment on Explorer, we need to consider a few scenarios here:

```
* Endpoint is given
  * And bucket is an empty string (s3:///path): keep bucket nil
  * Otherwise keep both endpoint and bucket in config
* Endpoint is nil
  * Bucket has dots: keep the endpoint _without_ the bucket, fill in bucket field
  * Bucket does not have dots: keep the endpoint with bucket, bucket is nil
```

This should allow us to use the bucket as a flag for virtual host or not, removing the need to check if endpoint contains the bucket.